src/shims/tls.rs

```diff
@@ -156,40 +189,63 @@ impl<'tcx> TlsData<'tcx> {
 
 impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
 pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
-    fn run_tls_dtors(&mut self) -> InterpResult<'tcx> {
+
+    /// Run TLS destructors for the main thread on Windows. The implementation
+    /// assumes that we do not support concurrency on Windows yet.
+    ///
+    /// Note: on non-Windows OS this function is a no-op.
+    fn run_windows_tls_dtors(&mut self) -> InterpResult<'tcx> {
         let this = self.eval_context_mut();
-        assert!(!this.machine.tls.dtors_running, "running TLS dtors twice");
-        this.machine.tls.dtors_running = true;
+        if this.tcx.sess.target.target.target_os != "windows" {
+            return Ok(());
+        }
+        let active_thread = this.get_active_thread()?;
+        assert_eq!(this.get_total_thread_count()?, 1, "concurrency on Windows not supported");
+        assert!(!this.machine.tls.dtors_running.contains(&active_thread), "running TLS dtors twice");
+        this.machine.tls.dtors_running.insert(active_thread);
+        // Windows has a special magic linker section that is run on certain events.
+        // Instead of searching for that section and supporting arbitrary hooks in there
+        // (that would be basically https://github.com/rust-lang/miri/issues/450),
+        // we specifically look up the static in libstd that we know is placed
+        // in that section.
+        let thread_callback = this.eval_path_scalar(&["std", "sys", "windows", "thread_local", "p_thread_callback"])?;
+        let thread_callback = this.memory.get_fn(thread_callback.not_undef()?)?.as_instance()?;
+
+        // The signature of this function is `unsafe extern "system" fn(h: c::LPVOID, dwReason: c::DWORD, pv: c::LPVOID)`.
+        let reason = this.eval_path_scalar(&["std", "sys", "windows", "c", "DLL_PROCESS_DETACH"])?;
+        let ret_place = MPlaceTy::dangling(this.machine.layouts.unit, this).into();
+        this.call_function(
+            thread_callback,
+            &[Scalar::null_ptr(this).into(), reason.into(), Scalar::null_ptr(this).into()],
+            Some(ret_place),
+            StackPopCleanup::None { cleanup: true },
+        )?;
+
+        // step until out of stackframes
+        this.run()?;
+
+        // Windows doesn't have other destructors.
+        Ok(())
+    }
 
+    /// Run TLS destructors for the active thread.
+    ///
+    /// Note: on Windows OS this function is a no-op because we do not support
+    /// concurrency on Windows yet.
+    ///
+    /// FIXME: we do not support yet deallocation of thread local statics.
```

Comment:
What does this last line mean? We run dtors, and we have a test checking that accessing thread-local statics after thread exit is UB, right?